overlord/assertstate/bulk.go

```diff
@@ -139,6 +140,62 @@ func bulkRefreshSnapDeclarations(s *state.State, snapStates map[string]*snapstat
 	return nil
 }
 
+func bulkRefreshValidationSetAsserts(s *state.State, vsets map[string]*ValidationSetTracking, userID int, deviceCtx snapstate.DeviceContext) error {
+	db := cachedDB(s)
+	pool := asserts.NewPool(db, maxGroups)
+
+	ignoreNotFound := make(map[string]bool)
+
+	for _, vs := range vsets {
+		var atSeq *asserts.AtSequence
+		if vs.PinnedAt > 0 {
+			// pinned to specific sequence, update to latest revision for same
+			// sequence.
+			atSeq = &asserts.AtSequence{
+				Type:        asserts.ValidationSetType,
+				SequenceKey: []string{release.Series, vs.AccountID, vs.Name},
+				Sequence:    vs.PinnedAt,
+				Pinned:      true,
+			}
+		} else {
+			// not pinned, update to latest sequence
+			atSeq = &asserts.AtSequence{
+				Type:        asserts.ValidationSetType,
+				SequenceKey: []string{release.Series, vs.AccountID, vs.Name},
+				Sequence:    vs.Current,
+			}
+		}
+		// every sequence to resolve has own group
+		grouping := atSeq.Unique()
+		if vs.LocalOnly {
+			ignoreNotFound[grouping] = true
+		}
+		if err := pool.AddSequenceToUpdate(atSeq, grouping); err != nil {
+			return err
+		}
+	}
+
+	err := resolvePool(s, pool, userID, deviceCtx)
```

Comment:
we need to do something about the bulkAssertionFallbackError we get here, because their current message is meant for logging and not for surfacing to the user but we don't need/can't fallback for valdation-sets

Reply:
Done.